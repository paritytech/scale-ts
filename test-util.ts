/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
