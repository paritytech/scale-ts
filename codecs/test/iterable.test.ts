import * as $ from "../../mod.ts"
import { testCodec } from "../../test-util.ts"

const $iterableArray = $.withMetadata(
  $.metadata("$iterableArray"),
  $.iterable<number, number[]>({
    $el: $.u8,
    calcLength: (arr) => arr.length,
    rehydrate: (iter) => [...iter],
    assert() {},
  }),
)

testCodec($iterableArray, [
  [],
  [0, 2, 4, 8],
  [2, 3, 5, 7],
])
