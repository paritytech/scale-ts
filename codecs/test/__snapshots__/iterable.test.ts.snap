export const snapshot = {};

snapshot[`\$.set(\$.u8) Set(0) {} 1`] = `00`;

snapshot[`\$.set(\$.u8) Set(4) { 0, 2, 4, 8 } 1`] = `
10
00
02
04
08
`;

snapshot[`\$.set(\$.u8) Set(4) { 2, 3, 5, 7 } 1`] = `
10
02
03
05
07
`;

snapshot[`\$.set(\$.u8) invalid null 1`] = `ScaleAssertError: !(value instanceof Set)`;

snapshot[`\$.set(\$.u8) invalid undefined 1`] = `ScaleAssertError: !(value instanceof Set)`;

snapshot[`\$.set(\$.u8) invalid 123 1`] = `ScaleAssertError: !(value instanceof Set)`;

snapshot[`\$.set(\$.u8) invalid [ 123 ] 1`] = `ScaleAssertError: !(value instanceof Set)`;

snapshot[`\$.set(\$.u8) invalid Set(1) { null } 1`] = `ScaleAssertError: typeof #iterator[0] !== "number"`;

snapshot[`\$.set(\$.u8) invalid Set(5) { 1, 2, 3, -1, 4 } 1`] = `ScaleAssertError: #iterator[3] < 0`;

snapshot[`\$.map(\$.str, \$.u8) Map(0) {} 1`] = `00`;

snapshot[`\$.map(\$.str, \$.u8) Map(2) { "0" => 0, "1" => 1 } 1`] = `
08
04
30
00
04
31
01
`;

snapshot[`\$.map(\$.str, \$.u8) Map(5) { "2^0" => 0, "2^1" => 2, "2^2" => 4, "2^3" => 8, "2^4" => 16 } 1`] = `
14
0c
32
5e
30
00
0c
32
5e
31
02
0c
32
5e
32
04
0c
32
5e
33
08
0c
32
5e
34
10
`;

snapshot[`\$.map(\$.str, \$.u8) invalid null 1`] = `ScaleAssertError: !(value instanceof Map)`;

snapshot[`\$.map(\$.str, \$.u8) invalid undefined 1`] = `ScaleAssertError: !(value instanceof Map)`;

snapshot[`\$.map(\$.str, \$.u8) invalid 123 1`] = `ScaleAssertError: !(value instanceof Map)`;

snapshot[`\$.map(\$.str, \$.u8) invalid [ 123 ] 1`] = `ScaleAssertError: !(value instanceof Map)`;

snapshot[`\$.map(\$.str, \$.u8) invalid [ [ "a", 1 ] ] 1`] = `ScaleAssertError: !(value instanceof Map)`;

snapshot[`\$.map(\$.str, \$.u8) invalid Map(1) { "a" => null } 1`] = `ScaleAssertError: typeof #iterator[0][1] !== "number"`;

snapshot[`\$.map(\$.str, \$.u8) invalid Map(4) { "a" => 1, "b" => 2, "c" => -1, "d" => 0 } 1`] = `ScaleAssertError: #iterator[2][1] < 0`;

snapshot[`\$.map(\$.str, \$.u8) invalid Map(4) { "a" => 1, "b" => 2, null => 3, "d" => 0 } 1`] = `ScaleAssertError: typeof #iterator[2][0] !== "string"`;

snapshot[`\$iterableArray [] 1`] = `00`;

snapshot[`\$iterableArray [ 0, 2, 4, 8 ] 1`] = `
10
00
02
04
08
`;

snapshot[`\$iterableArray [ 2, 3, 5, 7 ] 1`] = `
10
02
03
05
07
`;
